transport/tlsfrag/buffer.go

```diff
@@ -0,0 +1,153 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsfrag
+
+import (
+	"errors"
+	"fmt"
+	"io"
+)
+
+var (
+	// errTLSClientHelloFullyReceived is returned when a full TLS Client Hello has been received and no
+	// more data can be pushed to the buffer.
+	errTLSClientHelloFullyReceived = errors.New("already received a complete TLS Client Hello packet")
+
+	// errInvalidTLSClientHello is the error used when the data received is not a valid TLS Client Hello.
+	// Please use [errors.Is] to compare the returned err object with this instance.
+	errInvalidTLSClientHello = errors.New("not a valid TLS Client Hello packet")
+)
+
+// clientHelloBuffer is a byte buffer used to receive and buffer a TLS Client Hello packet.
+type clientHelloBuffer struct {
+	data   []byte // the buffer that hosts both header and content, len: 5 -> 5+len(content)
+	len    int    // the actual bytes that have been read into data
+	valid  bool   // indicate whether the content in data is a valid TLS Client Hello record
+	toRead int    // the number of bytes to read next, e.g. 5 -> len(content)
+}
+
+var _ io.Writer = (*clientHelloBuffer)(nil)
+var _ io.ReaderFrom = (*clientHelloBuffer)(nil)
+
+// newClientHelloBuffer creates and initializes a new buffer to receive a TLS Client Hello packet.
+func newClientHelloBuffer() *clientHelloBuffer {
+	// Allocate the 5 bytes header first, and then reallocate it to contain the entire packet later
+	return &clientHelloBuffer{
+		data:   make([]byte, recordHeaderLen),
+		valid:  true,
+		toRead: recordHeaderLen,
+	}
+}
+
+// Len returns the length of this buffer including both the 5 bytes header and the content.
+func (b *clientHelloBuffer) Len() int {
+	return b.len
+}
+
+// Bytes returns the full Client Hello packet including both the 5 bytes header and the content.
+func (b *clientHelloBuffer) Bytes() []byte {
+	return b.data[:b.len]
+}
+
+// Write appends p to the buffer and returns the number of bytes actually used.
+// If this data completes a valid TLS Client Hello, it returns errTLSClientHelloFullyReceived.
+// If an invalid TLS Client Hello message is detected, it returns the error errInvalidTLSClientHello.
+// If all bytes in p have been used and the buffer still requires more data to build a complete TLS Client Hello
+// message, it returns (len(p), nil).
+func (b *clientHelloBuffer) Write(p []byte) (n int, err error) {
+	if !b.valid {
+		return 0, errInvalidTLSClientHello
+	}
+
+	for b.len < len(b.data) && len(p) > 0 {
+		m := copy(b.data[b.len:], p)
+		n += m
+		b.len += m
+		p = p[m:]
+
+		if b.len == recordHeaderLen {
+			if err = b.validateTLSClientHello(); err != nil {
+				return
+			}
+			buf := make([]byte, recordHeaderLen+getMsgLen(b.data))
```

Comment:
We can implement that. But just note that it won't completely solve the timeout or write alignment issue:

* **Timeout** user can `Write` 4 valid bytes to us, and then `Read`, we will wait the 5th byte forever because we need to that to determine the total size.
* **Write Alignment** To solve it, we need additional allocation: user `split a 15 bytes (containing 10 bytes message) packet into three writes `Write(7)` `Write(5)` `Write(3)`, suppose we have a fixed 2 bytes split point, we will forward: `Write(7)` (with modified the new msgsize = 2), `Write(5)` (the new header with msgsize=8), `Write(5)` and than `Write(3)`. There is an additional `Write(5)` there. To merge it with the following content, we need to allocate a 10 bytes buffer to put both the header and the content.